tests/unit/dense_fold.cpp

```diff
@@ -0,0 +1,224 @@
+
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+
+#include <alp.hpp>
+
+using namespace alp;
+
+typedef double T1;
+
+const T1 testval1 = 1.5;
+const T1 testval2 = -1;
+const T1 testval3 = 2.;
+
+template< typename VectorType >
+void print_vector( std::string name, const VectorType &v ) {
+
+	if( ! alp::internal::getInitialized( v ) ) {
+		std::cout << "Vector " << name << " uninitialized.\n";
+		return;
+	}
+
+	std::cout << "Vector " << name << " of size " << alp::getLength( v ) << " contains the following elements:\n";
+
+	std::cout << "[\t";
+	for( size_t i = 0; i < alp::getLength( v ); ++i ) {
+		std::cout << v[ i ] << "\t";
+	}
+	std::cout << "]\n";
+}
+
+void alp_program( const size_t &n, alp::RC &rc ) {
+
+	// test 1 foldl( vector, scalar, mul_op)
+	{
+		rc = SUCCESS;
+
+		alp::Vector< T1 > x( n );
+
+		alp::Semiring<
+			alp::operators::add< double >, alp::operators::mul< double >,
+			alp::identities::zero, alp::identities::one
+			> ring;
+
+		std::vector< T1 > x_data( n );
+
+		// test 1, init
+		std::fill(x_data.begin(), x_data.end(), testval1 );
+
+		rc = rc ? rc : alp::buildVector( x, x_data.begin(), x_data.end() );
+		if( rc != SUCCESS ) {
+			std::cerr << "\t test 1 (foldl( vector, scalar, mul_op)): initialisation FAILED\n";
+			return;
+		}
+
+		// test 1, exec
+		Scalar< T1 > out( testval2 );
+		auto OP = ring.getMultiplicativeOperator();
```

Comment:
Either use `const auto` or pass `ring.getMultiplicativeOperator()` directly to the function call below.